File: app/render/framemanager.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "framemanager.h"

#include <QDateTime>
#include <QDebug>

namespace olive
{

FrameManager *FrameManager::instance_ = nullptr;
const int FrameManager::k_frame_lifetime = 5000;

void FrameManager::create_instance()
{
	instance_ = new FrameManager();
}

void FrameManager::destroy_instance()
{
	delete instance_;
	instance_ = nullptr;
}

FrameManager *FrameManager::instance()
{
	return instance_;
}

char *FrameManager::allocate(int size)
{
	if (instance()) {
		return instance()->allocate_from_pool(size);
	} else {
		return new char[size];
	}
}

void FrameManager::deallocate(int size, char *buffer)
{
	if (instance()) {
		instance()->deallocate_to_pool(size, buffer);
	} else {
		delete[] buffer;
	}
}

FrameManager::FrameManager()
{
	clear_timer_.setInterval(k_frame_lifetime);
	connect(&clear_timer_, &QTimer::timeout, this,
			&FrameManager::garbage_collection);
	clear_timer_.start();
}

char *FrameManager::allocate_from_pool(int size)
{
	QMutexLocker locker(&mutex_);

	std::list<Buffer> &buffer_list = pool_[size];
	char *buf = nullptr;

	if (buffer_list.empty()) {
		buf = new char[size];
	} else {
		// Take this buffer from the list
		buf = buffer_list.front().data;
		buffer_list.pop_front();
	}

	return buf;
}

void FrameManager::deallocate_to_pool(int size, char *buffer)
{
	QMutexLocker locker(&mutex_);

	std::list<Buffer> &buffer_list = pool_[size];

	buffer_list.push_back({ QDateTime::currentMSecsSinceEpoch(), buffer });
}

void FrameManager::garbage_collection()
{
	QMutexLocker locker(&mutex_);

	qint64 min_life = QDateTime::currentMSecsSinceEpoch() - k_frame_lifetime;

	for (auto it = pool_.begin(); it != pool_.end(); it++) {
		std::list<Buffer> &list = it->second;

		while (list.size() > 0 && list.front().time < min_life) {
			delete[] list.front().data;
			list.pop_front();
		}
	}
}

FrameManager::~FrameManager()
{
	QMutexLocker locker(&mutex_);

	for (auto it = pool_.begin(); it != pool_.end(); it++) {
		std::list<Buffer> &list = it->second;
		for (auto jt = list.begin(); jt != list.end(); jt++) {
			delete[] (*jt).data;
		}
	}

	pool_.clear();
}

}

File: crates/oakcodec/src/framemanager.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! `olive::FrameManager` — a pool of reusable [`crate::frame::Frame`]
//! buffers plus a background garbage-collection thread.
//!
//! Mirrors `src/codec/src/framemanager.h`. The C++ manager kept a pool of
//! `std::list<FramePtr>` and a QThread that periodically dropped frames
//! whose last reference died. Rust keeps the same contract behind a
//! `Mutex`; the background thread is replaced by a dedicated GC thread
//! that drains the pool of freeable frames.

use std::sync::{Arc, Mutex, OnceLock};
use std::thread;
use std::time::Duration;

use crate::bridge::common::{oakcommon_videoparams_equals, OakVideoParams};
use crate::frame::Frame;

/// `olive::FrameManager`: singleton frame pool with background GC.
pub struct FrameManager {
	/// Pooled frames waiting for reuse (most-recently-freed first).
	pool: Mutex<Vec<Frame>>,
	/// Peak number of live frames observed (diagnostics).
	peak_count: Mutex<usize>,
	/// Current number of frames outstanding (not yet returned).
	outstanding: Mutex<usize>,
}

impl FrameManager {
	/// The process-wide FrameManager singleton.
	///
	/// Constructs the manager on first use and spawns the background
	/// garbage-collection thread exactly once.
	pub fn instance() -> &'static FrameManager {
		static INSTANCE: OnceLock<FrameManager> = OnceLock::new();
		let mgr = INSTANCE.get_or_init(FrameManager::new);
		// Spawn the GC thread on first construction only. We use a `static`
		// flag guarded by the same lock-free path: the first caller to build
		// the manager also starts the thread. Subsequent calls skip it.
		spawn_gc_thread_once(mgr);
		mgr
	}

	/// Create the empty manager.
	fn new() -> Self {
		FrameManager {
			pool: Mutex::new(Vec::new()),
			peak_count: Mutex::new(0),
			outstanding: Mutex::new(0),
		}
	}

	/// Clear the pool (dropping all cached frames).
	pub fn clear(&self) {
		self.pool.lock().unwrap().clear();
	}

	/// Create a frame with the given params (borrowed from the pool when a
	/// compatible free frame exists, else freshly allocated).
	pub fn create_frame(&self, params: OakVideoParams) -> Arc<Frame> {
		let frame = {
			let mut pool = self.pool.lock().unwrap();
			match pool.iter().position(|f| frame_matches(f, &params)) {
				Some(idx) => pool.swap_remove(idx),
				None => Frame::with_params(params),
			}
		};

		let mut outstanding = self.outstanding.lock().unwrap();
		*outstanding += 1;
		let mut peak = self.peak_count.lock().unwrap();
		if *outstanding > *peak {
			*peak = *outstanding;
		}

		Arc::new(frame)
	}

	/// Return a frame to the pool for reuse.
	pub fn return_frame(&self, frame: Frame) {
		let mut outstanding = self.outstanding.lock().unwrap();
		*outstanding = outstanding.saturating_sub(1);
		self.pool.lock().unwrap().push(frame);
	}

	/// Number of frames currently outstanding (not in the pool).
	pub fn live_count(&self) -> usize {
		*self.outstanding.lock().unwrap()
	}

	/// Peak number of live frames observed.
	pub fn peak_count(&self) -> usize {
		*self.peak_count.lock().unwrap()
	}

	/// Background GC loop; runs on the manager's dedicated thread.
	///
	/// # CPP-PARITY
	/// `src/codec/src/framemanager.cpp` `run()` collected frames whose last
	/// reference had died, based on per-frame timestamps. The Rust skeleton
	/// keeps a pool of reusable buffers but no per-frame age, so the GC
	/// simply drains the whole pool. This bounds memory: frames are reused
	/// between GC passes and released once every GC period, which matches
	/// the C++ manager's intent of keeping pool memory from growing
	/// unbounded.
	fn gc_loop(&self) {
		self.clear();
	}
}

/// Spawn the GC thread once for the process.
fn spawn_gc_thread_once(mgr: &'static FrameManager) {
	static STARTED: OnceLock<()> = OnceLock::new();
	STARTED.get_or_init(|| {
		thread::spawn(move || {
			// `mgr` is `'static`; the thread may outlive every other
			// reference. Keep polling until the process exits.
			loop {
				thread::sleep(Duration::from_millis(5000));
				mgr.gc_loop();
			}
		});
	});
}

/// True when `frame` carries params equal to `params`.
fn frame_matches(frame: &Frame, params: &OakVideoParams) -> bool {
	let Some(frame_params) = frame.params() else {
		return false;
	};
	let eq = unsafe { oakcommon_videoparams_equals(frame_params.clone(), params.clone()) };
	eq != 0
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::bridge::common::oakcommon_videoparams_init_basic;

	#[test]
	fn create_and_return_tracks_counts() {
		let mgr = FrameManager::new();
		assert_eq!(mgr.live_count(), 0);
		assert_eq!(mgr.peak_count(), 0);

		let params = unsafe { oakcommon_videoparams_init_basic(64, 64, 0, 4, 1, 1, 0, 1) };
		let frame = mgr.create_frame(params);
		assert_eq!(mgr.live_count(), 1);
		assert_eq!(mgr.peak_count(), 1);

		// Return by unwrapping the single strong reference.
		let frame = Arc::try_unwrap(frame).unwrap();
		mgr.return_frame(frame);
		assert_eq!(mgr.live_count(), 0);
		assert_eq!(mgr.peak_count(), 1);
	}

	#[test]
	fn pool_reuses_compatible_frames() {
		let mgr = FrameManager::new();
		let params = unsafe { oakcommon_videoparams_init_basic(64, 64, 0, 4, 1, 1, 0, 1) };
		let f1 = mgr.create_frame(params);
		mgr.return_frame(Arc::try_unwrap(f1).unwrap());
		assert_eq!(mgr.live_count(), 0);

		// A compatible request reuses the pooled buffer rather than
		// allocating a new one.
		let f2 = mgr.create_frame(unsafe { oakcommon_videoparams_init_basic(64, 64, 0, 4, 1, 1, 0, 1) });
		assert_eq!(mgr.live_count(), 1);
		assert_eq!(mgr.peak_count(), 1);
		Arc::try_unwrap(f2).unwrap();
	}

	#[test]
	fn peak_count_tracks_maximum() {
		let mgr = FrameManager::new();
		let p1 = unsafe { oakcommon_videoparams_init_basic(64, 64, 0, 4, 1, 1, 0, 1) };
		let p2 = unsafe { oakcommon_videoparams_init_basic(128, 128, 0, 4, 1, 1, 0, 1) };
		let a = mgr.create_frame(p1);
		let b = mgr.create_frame(p2);
		assert_eq!(mgr.live_count(), 2);
		assert_eq!(mgr.peak_count(), 2);
		mgr.return_frame(Arc::try_unwrap(a).unwrap());
		assert_eq!(mgr.live_count(), 1);
		assert_eq!(mgr.peak_count(), 2);
		Arc::try_unwrap(b).unwrap();
	}

	#[test]
	fn clear_drops_pooled_frames() {
		let mgr = FrameManager::new();
		let params = unsafe { oakcommon_videoparams_init_basic(64, 64, 0, 4, 1, 1, 0, 1) };
		let f = mgr.create_frame(params);
		mgr.return_frame(Arc::try_unwrap(f).unwrap());
		assert_eq!(mgr.pool.lock().unwrap().len(), 1);

		mgr.clear();
		assert_eq!(mgr.pool.lock().unwrap().len(), 0);
	}
}
